### PR TITLE
read events using Date Ranges

### PR DESCRIPTION
**Pull request by Baalmart**
# read events using Date Ranges 

**_WHAT DOES THIS PR DO?_**
- [x] From data management microservice, read a Date Range of measurements in JSON format. Both recent and most recent from the optional date query parameters. Two new endpoints created for this.

**_WHAT JIRA STORY/TASK IS RELATED TO THIS PR?_**
[PLAT-868]

**_WHAT IS THE LINK TO THE PR BRANCH_**
https://github.com/airqo-platform/AirQo-api/tree/en-data-mgt

**_HOW DO I TEST OUT THIS PR?_**
```
cd AirQo-api/src/data-mgt/node
npm install
npm run stage-mac
```
🚨 The [.env](https://docs.google.com/document/d/1gX94v2Yavz_reD1yvC4NjJ_90gTQV7EhfckWRObeUQI/edit?usp=sharing) has also been updated accordingly. 

🚨 Also note that you need to change all the Redis cache values in your **.env** to 0 (Zero) on your local machines just to ensure that you do not have to "wait" when carrying out tests.  Otherwise, you will keep on getting the same results and think that the endpoint is not working yet you are getting cached results!

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 

- [x] [retrieve device readings using date ranges](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/data-management#retrieve-device-readings)
- [x] [retrieve most RECENT device readings using date ranges](https://app.gitbook.com/o/-MCogOVGQqjUharJKchL/s/-MKqyQkcfs54GYi-q96G/data-management#retrieve-the-most-recent-readings)

**_IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?_**
N/A

**_SCREENSHOTS?_**
Screenshot 2021-12-07 at 16 52 07
Screenshot 2021-12-07 at 16 51 14

[PLAT-868]: https://airqoteam.atlassian.net/browse/PLAT-868?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by NoahNsimbe — COMMENTED** [review]
@Baalmart please provide support for querying by time.

**Comment by NoahNsimbe** [comment]
> * retrieve device readings

Hi @Baalmart , we still need to perform thorough tests. Querying by start and end time has issues. 

- [ ] Query by start time alone
- [ ] Query by end time alone
- [ ] Query by both start and end time

For all the above, we need to put into consideration the hours and minutes while testing.

**Comment by Baalmart** [comment]
> > * retrieve device readings
> 
> Hi @Baalmart , we still need to perform thorough tests. Querying by start and end time has issues.
> 
> * [ ]  Query by start time alone
> * [ ]  Query by end time alone
> * [ ]  Query by both start and end time
> 
> For all the above, we need to put into consideration the hours and minutes while testing.

Hi @NoahNsimbe , please provide more details of "the issues" so that I can act accordingly. Use screenshots where need be for me to easily address the issues.

Also, please note that the .env has just been updated so please check it out asap [here](https://docs.google.com/document/d/1gX94v2Yavz_reD1yvC4NjJ_90gTQV7EhfckWRObeUQI/edit?usp=sharing).

**Comment by NoahNsimbe** [comment]
> > > * retrieve device readings
> > 
> > 
> > Hi @Baalmart , we still need to perform thorough tests. Querying by start and end time has issues.
> > 
> > * [ ]   Query by start time alone
> > * [ ]   Query by end time alone
> > * [ ]   Query by both start and end time
> > 
> > For all the above, we need to put into consideration the hours and minutes while testing.
> 
> Hi @NoahNsimbe , please provide more details of "the issues" so that I can act accordingly. Use screenshots where need be for me to easily address the issues.
> 
> Also, please note that the .env has just been updated so please check it out asap [here](https://docs.google.com/document/d/1gX94v2Yavz_reD1yvC4NjJ_90gTQV7EhfckWRObeUQI/edit?usp=sharing).

Hi @Baalmart, you just have to make sure querying by `date time` functions as expected. The last time we checked, it only supported querying by `date`.  For example. If I specify `2022-01-01T10:00:00Z` and `2022-01-01T18:00:00Z` as my `start date time` and `end date time` resp, the endpoint should return events recorded on `2022-01-01`  between `10am` and `6pm`

**Review by NoahNsimbe — COMMENTED** [review]
Querying by date time ranges LGTM

**Review by NoahNsimbe — COMMENTED** [review]
@123MwanjeMike, there is a change in the `.env` You might want to update the configmaps of the service later.

**Review by 123MwanjeMike — APPROVED** [review]
Thanks @Baalmart 
LGTM

**Review by NoahNsimbe — APPROVED** [review]
LGTM

File: src/data-mgt/node/config/constants.js [hunk 1]
```diff
@@ -1,109 +1,141 @@
 const axios = require("axios");
 const isEmpty = require("is-empty");
 const { logElement, logText, logObject } = require("../utils/log");
-const jsonify = require("../utils/jsonify");
 
 const devConfig = {
   MONGO_URI: "mongodb://localhost/",
   JWT_SECRET: process.env.JWT_SECRET,
   DB_NAME: process.env.MONGO_DEV,
   REDIS_SERVER: process.env.REDIS_SERVER_DEV,
```
Comment (Baalmart) [inline src/data-mgt/node/config/constants.js:9]: Hi @123MwanjeMike , we need to first update the ENVs in both the production and staging clusters before we merge this PR.

The updated .env was shared in the PR description....... cc: @NoahNsimbe

File: src/data-mgt/node/config/constants.js [hunk 2]
```diff
@@ -1,109 +1,141 @@
...
-        return {
-          success: false,
-          error: error.message,
-          message: "constants server side error",
-        };
-      });
+
+  BAM_FIELDS_AND_DESCRIPTIONS: {
+    field1: "RTCADC",
+    field2: "RTCV",
```
Comment (Baalmart) [inline src/data-mgt/node/config/constants.js:91]: these are BAM config details...

File: src/data-mgt/node/config/redis.js [hunk 3]
```diff
@@ -1,18 +1,22 @@
...
 const REDIS_SERVER = constants.REDIS_SERVER;
 const REDIS_PORT = constants.REDIS_PORT;
 logElement("redis URL", REDIS_SERVER.concat(":", REDIS_PORT));
 
 const client = redis.createClient({
   host: REDIS_SERVER,
-  port: REDIS_PORT
-}
-);
+  port: REDIS_PORT,
```
Comment (Baalmart) [inline src/data-mgt/node/config/redis.js:10]: few adjustments to the config details of Redis

File: src/data-mgt/node/routes/api.js [hunk 4]
```diff
@@ -16,4 +17,36 @@ router.get(
   transformController.generateDescriptiveLastEntry
 );
 
+router.get(
+  "/feeds",
```
Comment (Baalmart) [inline src/data-mgt/node/routes/api.js:21]: A new endpoint for getting all measurements of provided device

File: src/data-mgt/node/routes/api.js [hunk 5]
```diff
@@ -16,4 +17,36 @@ router.get(
...
+      .bail()
+      .trim()
+      .isInt()
+      .withMessage("the channel should be an integer value"),
+  ]),
+  transformController.readFeeds
+);
+
+router.get(
+  "/feeds/last",
```
Comment (Baalmart) [inline src/data-mgt/node/routes/api.js:37]: A new endpoint for getting the most recent measurements of provided device

File: src/data-mgt/node/utils/date.js [hunk 6]
```diff
@@ -15,4 +15,138 @@ const generateDateFormat = async (ISODate) => {
   return `${year}-${month}-${dt}-${hrs}`;
 };
 
-module.exports = { generateDateFormat };
+const isTimeEmpty = (dateTime) => {
+  let date = new Date(dateTime);
+  let hrs = date.getUTCHours();
+  let mins = date.getUTCMinutes();
+  let secs = date.getUTCSeconds();
```
Comment (Baalmart) [inline src/data-mgt/node/utils/date.js:22]: introduced more functions to the date util

File: src/data-mgt/node/utils/errors.js [hunk 7]
```diff
@@ -1,56 +1,28 @@
...
+    return controller.reduce((obj, item) => {
+      let param = item.param;
+      return {
+        ...obj,
+        [param]: `${item.msg}`,
+      };
+    }, initialValue);
+  },
+  errorResponse: ({ res, message, statusCode = 500, error = {} } = {}) => {
+    logObject("error", error);
```
Comment (Baalmart) [inline src/data-mgt/node/utils/errors.js:16]: To emphasise the global error handling, merged all the functions into one for all error responses irrespective of the type

File: src/data-mgt/node/utils/transform.js [hunk 8]
```diff
@@ -0,0 +1,302 @@
+const axios = require("axios");
+const isEmpty = require("is-empty");
+const { logElement, logText, logObject } = require("./log");
+const constants = require("../config/constants");
```
Comment (Baalmart) [inline src/data-mgt/node/utils/transform.js:4]: merged all utils for measurement transformation into one file. As I prepare for the integration of unit tests to all files in the codebase.

**Review by Baalmart — COMMENTED** [review]
Hi @123MwanjeMike , could we update the ENVs for this micro in its respective clusters for both prod and stage.

Afterwards, please proceed with the merge duty. Thanks!

**Comment by 123MwanjeMike** [comment]
> Thanks @NoahNsimbe, the env has been updated

Hi @Baalmart, the environment variables in our clusters we updated yesterday. Could there be any more new ones?

**Pull request merged**